### PR TITLE
Support publishing steganographic posts in facebook and twitter

### PR DESCRIPTION
**Pull request by guanbinrui**
This PR includes:

+ add a new trigger in developer panel to enable steganographic publishing;
+ add `SteganographyService` for encoding and decoding steganographic posts;
+ auto downloading encoded image but manually uploading is required in facebook;
+ auto decoding all images founded in facebook timeline and append result into `postContent`;

Related Issues:

#304 #316

### CODE REVIEW

**Comment by yisiliu** [comment]
If merged, we will include this feature in our next release :)

**Comment by SunriseFox** [comment]
I have had the switch turned on... I could post with images but could not decrypt it.

**Comment by guanbinrui** [comment]
@SunriseFox Can you give the post URL? Let me check it.

**Comment by SunriseFox** [comment]
@guanbinrui Try this: https://www.facebook.com/photo.php?fbid=616286175573625&set=a.554179138450996&type=3

**Comment by guanbinrui** [comment]
> @guanbinrui Try this: https://www.facebook.com/photo.php?fbid=616286175573625&set=a.554179138450996&type=3

Facebook tells me the link may be broken.

File: src/extension/background-script/SteganographyService.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+const defaultOptions = {
+    size: 8,
+    narrow: 0,
+    copies: 3,
+    tolerance: 128,
+}
+
+export async function encodeImage(
+    { buffer }: Uint8Array,
```
Comment (Jack-Works) [inline src/extension/background-script/SteganographyService.ts]: Question: Does Uint8Array can be transported across the context?

File: src/social-network-provider/facebook.com/tasks/uploadToPostBox.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { SocialNetworkUI, getActivatedUI } from '../../../social-network/ui'
+import { untilDocumentReady } from '../../../utils/dom'
+import { getUrl, downloadUrl } from '../../../utils/utils'
+import Services from '../../../extension/service'
+
+export async function uploadToPostBoxFacebook(
+    this: SocialNetworkUI,
```
Comment (Jack-Works) [inline src/social-network-provider/facebook.com/tasks/uploadToPostBox.ts]: suggestion: avoid the use of `this: SocialNetworkUI`, this pattern has encountered some problems in previous development

File: src/extension/background-script/SteganographyService.ts (hunk 1 above)
Comment (guanbinrui) [reply src/extension/background-script/SteganographyService.ts]: Yes `Uint8Array` can persist transporting. I perfer use `ArrayBuffer` here but lost content after received in the other side.

Comment (Jack-Works) [reply src/extension/background-script/SteganographyService.ts]: 🤔Oh I found serialijse support TypedArray

**Comment by SunriseFox** [comment]
Facebook gives me too small images to decrypt from... I tried to clear cache, alter my user agents and use proxies but still got the same small image.

**Comment by yisiliu** [comment]
I'd like to wait until auto image posting is done before merging.

File: src/_locales/en/messages.json [hunk 3]
```diff
@@ -17,6 +17,12 @@
     "additional_post_box__encrypted_failed": {
         "message": "Please paste the following text into the post box."
     },
+    "additional_post_box__steganography_post_pre": {
+        "message": "This post is shared with my friends. You may install Maskbook to decrypt the content from the image. ([I:b])"
```
Comment (Artoria2e5) [inline src/_locales/en/messages.json]: > ([I:b])

?

Comment (guanbinrui) [reply src/_locales/en/messages.json]: Copy & Paste from #304

Comment (neruthes) [reply src/_locales/en/messages.json]: It is the yanwenzi for The Laughing Mask logo of Maskbook.

Comment (Jack-Works) [reply src/_locales/en/messages.json]: 絵文字？

Comment (neruthes) [reply src/_locales/en/messages.json]: 颜文字 is different from emoji. This practice uses existing characters to construct emotional expressions. Examples include `:)` and `XD`.

File: src/utils/utils.ts [hunk 4]
```diff
@@ -26,6 +35,45 @@ export function dispatchCustomEvents<T extends keyof CustomEvents>(event: T, ...
...
+    const textWrapper = document.createElement('script')
+    textWrapper.setAttribute('id', 'm-aaaaaaaa')
+    textWrapper.setAttribute('type', 'text/x-custom-script')
+    textWrapper.innerHTML = bytes.toString()
+    document.body.append(textWrapper)
+
+    const script = document.createElement('script')
+    script.setAttribute('async', 'false')
+    script.innerHTML = `
+        const s = document.body.querySelector('script#m-aaaaaaaa').innerHTML
```
Comment (Jack-Works) [inline src/utils/utils.ts]: 🤔🤔 You may extend injected script to archive your purpose

Comment (neruthes) [reply src/utils/utils.ts]: > `const s =`

Reduce bytes used? 🤔

Comment (SunriseFox) [reply src/utils/utils.ts]: I think this is better... A one-time generated script run and destroy, and CSP-proof. Works on iOS and twitter too.

Comment (SunriseFox) [reply src/utils/utils.ts]: ...and reduce megabytes used.

Comment (Jack-Works) [reply src/utils/utils.ts]: It's not safe to inject a script tag in the non-init time.

Comment (SunriseFox) [reply src/utils/utils.ts]: could you please explain it? I didn't see if it brings more vulnerabilities to our implementation.

**Pull request merged**